### PR TITLE
fix panic when the rpc client is not initialized

### PR DESCRIPTION
**Pull request by krasi-georgiev**
**What type of PR is this?**

Bug fix


fixes this panic when the code tries to call the rpc client before it is initialized

```
[2022-09-26 07:03:54]  INFO initial-sync: Processing block batch of size 25 starting from  0xb89eddc1... 3960641/3972319 - estimated time remaining 2h35m42s blocksPerSecond=1.2 
peers=6                                                                                                                                                                          
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                          
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x23abfa6]                                                                                                         
                                                                                                                                                                                 
goroutine 326 [running]:                                                                                                                                                         
github.com/prysmaticlabs/prysm/v3/beacon-chain/execution.(*Service).NewPayload(0xc000012380, {0xbbad20?, 0xc046a836e0?}, {0xbcba90, 0xc01e95b040})                               
        beacon-chain/execution/engine_client.go:100 +0x246                                                                                                                       
github.com/prysmaticlabs/prysm/v3/beacon-chain/blockchain.(*Service).notifyNewPayload(0xc0001d0000, {0xbbad20?, 0xc023134f90?}, 0x2, 0xbc9e40?, {0xbc9e40, 0xc01d3c1b90})        
        beacon-chain/blockchain/execution_engine.go:215 +0x227                                                                                                                   
github.com/prysmaticlabs/prysm/v3/beacon-chain/blockchain.(*Service).onBlockBatch(0xc0001d0000, {0xbbad20?, 0xc023134f60?}, {0xc01a3261c0, 0x19, 0x0?}, {0xc023eb1c00, 0x19, 0x0?
})                                                                                                                                                                               
        beacon-chain/blockchain/process_block.go:416 +0x1212                                                                                                                     
github.com/prysmaticlabs/prysm/v3/beacon-chain/blockchain.(*Service).ReceiveBlockBatch(0xc0001d0000, {0xbbac78?, 0xc01a127440?}, {0xc01a3261c0, 0x19, 0x24}, {0xc023eb1c00, 0x19,
 0x19})                                                                                                                                                                          
        beacon-chain/blockchain/receive_block.go:93 +0x151                                                                                                                       
github.com/prysmaticlabs/prysm/v3/beacon-chain/sync/initial-sync.(*Service).processBatchedBlocks(0xc0133a8240, {0xbbac78, 0xc01a127440}, {0x186eca5?, 0x60?, 0x30d9400?}, {0xc01a
326000, 0x35, 0x40}, 0xc00d371630)                                                                                                                                               
        beacon-chain/sync/initial-sync/round_robin.go:288 +0x8e7                                                                                                                 
github.com/prysmaticlabs/prysm/v3/beacon-chain/sync/initial-sync.(*Service).processFetchedData(0xc01a125b90?, {0xbbac78?, 0xc01a127440?}, {0xc00d371680?, 0x0?, 0x30d9400?}, 0xc0
1a119548?, 0x0?)                                                                                                                                                                 
        beacon-chain/sync/initial-sync/round_robin.go:131 +0x11e                                                                                                                 
github.com/prysmaticlabs/prysm/v3/beacon-chain/sync/initial-sync.(*Service).syncToFinalizedEpoch(0xc0133a8240, {0xbbac78, 0xc01a127440}, {0x337f60?, 0x2ec4a30?, 0x30d9400?})    
        beacon-chain/sync/initial-sync/round_robin.go:84 +0x285                                                                                                                  
github.com/prysmaticlabs/prysm/v3/beacon-chain/sync/initial-sync.(*Service).roundRobinSync(0xc0133a8240, {0xe00000004?, 0xc000087ce0?, 0x30d9400?})                              
        beacon-chain/sync/initial-sync/round_robin.go:47 +0x187                                                                                                                  
github.com/prysmaticlabs/prysm/v3/beacon-chain/sync/initial-sync.(*Service).Start(0xc0133a8240)                                                                                  
        beacon-chain/sync/initial-sync/service.go:105 +0x597                                                                                                                     
created by github.com/prysmaticlabs/prysm/v3/runtime.(*ServiceRegistry).StartAll                                                                                                 
        runtime/service_registry.go:46 +0x210       
```

also reported in the flashbot repo
https://github.com/flashbots/prysm/issues/7

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Interesting approach to fixing this. Our typical solution is to add nil checks, but I am wondering why this client was not set in the constructor anyway.

cc: @rauljordan (https://github.com/prysmaticlabs/prysm/pull/10498)

Raul, any thoughts here?

**Comment by nisdas** [comment]
@prestonvanloon The reason it isn't added is because we need to create an active connection with the execution client. So doing this in the constructor seemed wrong.

**Comment by krasi-georgiev** [comment]
What fix would you suggest?

**Comment by nisdas** [comment]
@krasi-georgiev , your approach is fine. I was just explaining to preston why we cant create it in the constructor

File: beacon-chain/execution/service.go [hunk 1]
```diff
@@ -112,6 +112,18 @@ type RPCClient interface {
 	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
 }
 
+type RPCClientUninitialize struct {
+}
+
+func (RPCClientUninitialize) Close() {}
+func (RPCClientUninitialize) BatchCall([]gethRPC.BatchElem) error {
+	return errors.New("BatchCall to a RPCClientUnitialized")
```
Comment (nisdas) [inline beacon-chain/execution/service.go:120]:
```suggestion
	return errors.New("rpc client is not initialized)
```

File: beacon-chain/execution/service.go [hunk 2]
```diff
@@ -112,6 +112,18 @@ type RPCClient interface {
...
+}
+
+func (RPCClientUninitialize) CallContext(context.Context, interface{}, string, ...interface{}) error {
+	return errors.New("CallContext to a RPCClientUnitialized")
```
Comment (nisdas) [inline beacon-chain/execution/service.go:124]:
```suggestion
	return errors.New("rpc client is not initialized")
```

File: beacon-chain/execution/service.go [hunk 3]
```diff
@@ -112,6 +112,18 @@ type RPCClient interface {
 	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
 }
 
+type RPCClientUninitialize struct {
```
Comment (nisdas) [inline beacon-chain/execution/service.go:115]:
```suggestion
type RPCClientEmpty struct {
```

**Review by nisdas — APPROVED** [review]

**Pull request merged**